From bce75577cf72285580f6aba02b10040467853c76 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Wed, 17 Apr 2024 18:40:58 -0400
Subject: [PATCH 3/9] Add pin-to-home button to connection details page

The dynamic app shortcuts list only shows the same 4 connections, and
it's not even frequency-based due to the way it gets rebuilt, so it's
kind of useless if you have more than 4 connections.

Updating:
- Add the resources.
- Find the onMenuItemClick method for the details toolbar (try searching for the value of @id/menu_delete_connection).
- Find the connection info class (it's used many times in the file -- it's usually the private field between two Uri fields), and diff it with the old one to ensure the method names are up-to-date.
- Also ensure resource IDs are up to date.
- Test to ensure add-to-home works.
- Test to ensure shortcuts update when changing the name.
- Test to ensure shortcuts open the correct connection.

diff --git a/res/menu/summary.xml b/res/menu/summary.xml
index 9d54522..f92ba3b 100644
--- a/res/menu/summary.xml
+++ b/res/menu/summary.xml
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <menu
   xmlns:android="http://schemas.android.com/apk/res/android" xmlns:app="http://schemas.android.com/apk/res-auto">
+    <item android:icon="@drawable/ic_home_white_24dp" android:id="@id/menu_home" android:orderInCategory="0" android:title="@string/menu_home_add" app:showAsAction="ifRoom" />
     <item android:icon="@drawable/ic_menu_delete" android:id="@id/menu_delete_connection" android:orderInCategory="0" android:title="@string/menu_delete_connection" app:showAsAction="ifRoom" />
     <item android:icon="@drawable/ic_help_outline_white_24dp" android:id="@id/menu_help" android:orderInCategory="0" android:title="@string/menu_help" app:showAsAction="collapseActionView" />
 </menu>
\ No newline at end of file
diff --git a/res/values/strings.xml b/res/values/strings.xml
index e5a5d51..6b7ba20 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -526,6 +526,7 @@ Your computers appear automatically; you no longer have to forward ports and con
     <string name="menu_hide_keyboard">Keyboard</string>
     <string name="menu_hide_mouse">Mouse buttons</string>
     <string name="menu_home">Home</string>
+    <string name="menu_home_add">Add to home screen</string>
     <string name="menu_info">Info</string>
     <string name="menu_no_filter">No filter</string>
     <string name="menu_offline">Only offline devices</string>
diff --git a/smali/com/realvnc/viewer/android/app/g3.smali b/smali/com/realvnc/viewer/android/app/g3.smali
index 97d906f..e0c929c 100644
--- a/smali/com/realvnc/viewer/android/app/g3.smali
+++ b/smali/com/realvnc/viewer/android/app/g3.smali
@@ -2043,6 +2043,10 @@
     const v0, 0x7f090216
 
     if-eq p1, v0, :cond_0
+ 
+    const v0, 0x7f090217 # @id/menu_home
+
+    if-eq p1, v0, :cond_4
 
     const/4 p1, 0x0
 
@@ -2113,9 +2117,67 @@
 
     :cond_3
     invoke-direct {p0}, Lcom/realvnc/viewer/android/app/g3;->k1()V
+ 
+    :cond_4
+    invoke-direct {p0}, Lcom/realvnc/viewer/android/app/g3;->addToHomeScreen()V
 
     return v1
 .end method
+ 
+.method private addToHomeScreen()V
+    .locals 5
+
+    invoke-virtual {p0}, Lcom/realvnc/viewer/android/app/g3;->A()Lcom/realvnc/viewer/android/app/ExtendedActivity;
+    move-result-object v1
+
+    # note that all shortcuts get rebuilt and enabled/disabled frequently by
+    # w2.v1 (search disableShortcuts), so this should preferably match that
+    # logic
+
+    iget-object v2, p0, Lcom/realvnc/viewer/android/app/g3;->Y:Lw2/e;
+    invoke-virtual {v2}, Lw2/e;->E()Ljava/lang/String; # connection uuid
+    move-result-object v2
+
+    new-instance v0, Landroid/content/pm/ShortcutInfo$Builder;
+    invoke-direct {v0, v1, v2}, Landroid/content/pm/ShortcutInfo$Builder;-><init>(Landroid/content/Context;Ljava/lang/String;)V
+
+    iget-object v2, p0, Lcom/realvnc/viewer/android/app/g3;->Y:Lw2/e;
+    invoke-virtual {v2}, Lw2/e;->A()Ljava/lang/String; # name or url
+    move-result-object v2
+    invoke-virtual {v0, v2}, Landroid/content/pm/ShortcutInfo$Builder;->setShortLabel(Ljava/lang/CharSequence;)Landroid/content/pm/ShortcutInfo$Builder;
+    move-result-object v0
+
+    const v2, 0x7f08016e # @drawable/preview_thumb_list
+    invoke-static {v1, v2}, Landroid/graphics/drawable/Icon;->createWithResource(Landroid/content/Context;I)Landroid/graphics/drawable/Icon;
+    move-result-object v2
+    invoke-virtual {v0, v2}, Landroid/content/pm/ShortcutInfo$Builder;->setIcon(Landroid/graphics/drawable/Icon;)Landroid/content/pm/ShortcutInfo$Builder;
+    move-result-object v0
+
+    iget-object v2, p0, Lcom/realvnc/viewer/android/app/g3;->Y:Lw2/e;
+    invoke-virtual {v2}, Lw2/e;->E()Ljava/lang/String;
+    move-result-object v2
+    invoke-static {v2}, Lw2/e;->C(Ljava/lang/String;)Landroid/net/Uri; # content://com.realvnc.viewer.android.data.AddressBook/entries/UUID
+    move-result-object v2
+    const/4 v3, 0x1 # is a shortcut
+    const/4 v4, 0x1 # shortcut is pinned
+    invoke-static {v2, v3, v4}, Lcom/realvnc/viewer/android/app/Application;->a(Landroid/net/Uri;ZZ)Landroid/content/Intent;
+    move-result-object v2
+    invoke-virtual {v0, v2}, Landroid/content/pm/ShortcutInfo$Builder;->setIntent(Landroid/content/Intent;)Landroid/content/pm/ShortcutInfo$Builder;
+    move-result-object v0
+
+    invoke-virtual {v0}, Landroid/content/pm/ShortcutInfo$Builder;->build()Landroid/content/pm/ShortcutInfo;
+    move-result-object v0
+
+    const-string v2, "shortcut"
+    invoke-virtual {v1, v2}, Landroid/content/Context;->getSystemService(Ljava/lang/String;)Ljava/lang/Object;
+    move-result-object v2
+    check-cast v2, Landroid/content/pm/ShortcutManager;
+
+    const/4 v3, 0x0
+    invoke-virtual {v2, v0, v3}, Landroid/content/pm/ShortcutManager;->requestPinShortcut(Landroid/content/pm/ShortcutInfo;Landroid/content/IntentSender;)Z
+
+    return-void
+.end method
 
 .method public final p(Ljava/lang/String;)V
     .locals 0
